### DISCUSSION TITLE
Managing GitHub Enterprise/Organisations/groups/repos with Infrastructure as a Code (IaaC) ?

### DISCUSSION THREAD

**Discussion by przemolb** [community/community#59916] [Enterprise]
### Select Topic Area

Question

### Body

Hello,

is there any way to store GitHub Enterprise Organizations, groups and repos as IaaC ? I know there is a Terraform provider but AFAIK it is way behind what is possible using i.e. GUI.
Any hint how can we manage all the structure using IaaC ?

**Comment by mateothegreat** [comment]
I've successfully implemented the lifecycle of our GitHub repositories with Terraform (now that the kinks have been worked out in the latest provider).

You'll find that you have almost 99% coverage on creation, branch protection rules, etc.

I've gone as far as to wrap change control around a "config" repo that serves as the data source.

Good luck!

**Reply by sunshinesissy** [reply]
![Screenshot_20231227_164500_Shop Samsung](https://github.com/community/community/assets/154071353/3d22104b-01f3-436f-b4bb-d06512dd37e3)

**Comment by przemolb** [comment]
The repositories is the easiest part.
What about organisations, policies, rulesets ?

**Reply by przemolb** [reply]
Any idea how organisations or rulesets are supported in Terraform ?

**Reply by przemolb** [reply]
Trying to refresh this thread and find people who manages GH organisations and rules using TF.

**Comment by taofikkhris** [comment]
Managing GitHub Enterprise or organizations, groups, and repositories with Infrastructure as Code (IaaC) is a powerful approach to automate the setup and configuration of your GitHub environment. Infrastructure as Code allows you to define and manage your GitHub resources using version-controlled code, making it easier to maintain, reproduce, and scale your GitHub infrastructure. Here's how you can use IaaC to manage various aspects of GitHub:

GitHub Enterprise Setup: If you are using GitHub Enterprise, you can use IaaC tools like Terraform or AWS CloudFormation to provision and configure your GitHub Enterprise instance. These tools can help you define the desired state of your GitHub Enterprise environment, including hardware specifications, network settings, and other configurations.

Organizations and Groups: With IaaC, you can create and manage GitHub organizations and groups programmatically. Using tools like GitHub's REST API or GitHub CLI, you can define the organization structure, set up teams and permissions, and invite members automatically.

Repositories: IaaC allows you to define your repositories' structure and settings as code. For example, you can use GitHub's API or CLI to create repositories, set default branch protections, enable certain features, and manage webhooks.

Access Controls: IaaC enables you to automate access controls across your GitHub repositories. You can define who has read or write access to specific repositories, teams, or organizations using code, making it easier to maintain consistent security practices.

GitHub Actions Workflows: GitHub Actions can also be defined as code. With IaaC, you can set up and manage workflows programmatically, defining how your repositories are built, tested, and deployed automatically based on specific triggers or events.

Security Settings: IaaC allows you to manage security-related settings on your GitHub repositories, such as enabling code scanning, dependency scanning, and secret scanning.

Monitoring and Auditing: By defining your GitHub infrastructure as code, you can easily track changes, maintain version history, and audit any modifications made to your GitHub environment.

Popular IaaC tools like Terraform, AWS CloudFormation, or GitHub's own API and CLI can be used to implement these automation tasks. By using IaaC, you can ensure consistency, repeatability, and scalability in managing your GitHub Enterprise, organizations, groups, and repositories, reducing manual efforts and potential errors in the process.

**Comment by mvkaran** [comment]
Hey @przemolb 👋🏼 

I am assuming you are looking for a way to manage **membership** of GitHub Enterprise orgs, teams and repos using IaC. One of the options you can explore, outside of Terraform, would be a Ruby gem named [entitlements-app](https://github.com/github/entitlements-app) that we at GitHub have open-sourced. This, in conjunction with a [GitHub plugin for it](https://github.com/github/entitlements-github-plugin) can help manage organization and team membership using GitOps.

**Reply by przemolb** [reply]
> Hey @przemolb 👋🏼 
> 
> I am assuming you are looking for a way to manage **membership** of GitHub Enterprise orgs, teams and repos using IaC. One of the options you can explore, outside of Terraform, would be a Ruby gem named [entitlements-app](https://github.com/github/entitlements-app) that we at GitHub have open-sourced. This, in conjunction with a [GitHub plugin for it](https://github.com/github/entitlements-github-plugin) can help manage organization and team membership using GitOps.

Hi @mv,

Thanks for the suggestion.
Unfortunately we don't use Ruby and we are Terraform shop.
My question is about the maturity of the TF GitHub provider and whether it covers all GH objects.

**Reply by mvkaran** [reply]
@przemolb Got it. The Terraform provider is being well used and has most of the objects supported through the API. If there's anything that you find missing, please do let us know!

**Reply by timme04** [reply]
Is there something else than the terraform provider? 
I like the way entitlements seems to manage assignments for changes to the ones who are responsible, but unfortunately it feels limited, for managing the actual organization aside from users and role assignments.

I would love to see options for creating actual repositories, and configuring settings on the repo.
